File: compute/pyo3/python/mog/sub_apis/hyperlinks.py
"""Hyperlink operations -- ``ws.hyperlinks.set()``, ``ws.hyperlinks.get()``."""
from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any, Dict, Optional, Tuple, Union

from mog._serde import deserialize_mutation_result, parse_a1
from mog.types import MutationResult

if TYPE_CHECKING:
    from mog._bridge import Bridge


class HyperlinksAPI:
    """Hyperlink operations on a worksheet."""

    __slots__ = ("_bridge", "_sheet_id_json")

    def __init__(self, bridge: Bridge, sheet_id_json: str) -> None:
        self._bridge = bridge
        self._sheet_id_json = sheet_id_json

    @staticmethod
    def _resolve_address(address: Union[str, Tuple[int, int]]) -> Tuple[int, int]:
        if isinstance(address, tuple):
            return address
        return parse_a1(address)

    def set(
        self,
        address: Union[str, Tuple[int, int]],
        url: str,
        display: Optional[str] = None,
        tooltip: Optional[str] = None,
    ) -> MutationResult:
        """Set a hyperlink on a cell.

        Parameters
        ----------
        address:
            A1 address or ``(row, col)`` tuple.
        url:
            The hyperlink URL or reference.
        display:
            Optional display text (defaults to the URL).
        tooltip:
            Optional tooltip text.
        """
        row, col = self._resolve_address(address)
        link = {"url": url}
        if display is not None:
            link["display"] = display
        if tooltip is not None:
            link["tooltip"] = tooltip
        raw = self._bridge.call_json(
            "compute_set_hyperlink",
            self._sheet_id_json,
            row,
            col,
            json.dumps(link),
        )
        return deserialize_mutation_result(raw)

    def get(self, address: Union[str, Tuple[int, int]]) -> Optional[str]:
        """Get the hyperlink URL at a cell, or ``None`` if none exists.

        Parameters
        ----------
        address:
            A1 address or ``(row, col)`` tuple.

        Returns the URL string, or ``None``.

        .. note::
            This matches the TypeScript SDK which returns ``string | null``.
            Use :meth:`get_full` to get the full dict with ``url``, ``display``,
            ``tooltip`` keys.
        """
        row, col = self._resolve_address(address)
        result = self._bridge.call_json(
            "compute_get_hyperlink", self._sheet_id_json, row, col
        )
        if isinstance(result, dict):
            url = result.get("url") or result.get("target")
            return url if url else None
        # Handle double-encoded JSON strings
        if isinstance(result, str) and result.startswith("{"):
            try:
                parsed = json.loads(result)
                if isinstance(parsed, dict):
                    url = parsed.get("url") or parsed.get("target")
                    return url if url else None
            except (json.JSONDecodeError, TypeError):
                pass
        if isinstance(result, str) and result:
            return result
        return None

    def get_full(self, address: Union[str, Tuple[int, int]]) -> Optional[Dict[str, Any]]:
        """Get the full hyperlink dict at a cell, or ``None`` if none exists.

        Returns a dict with ``url``, ``display``, ``tooltip`` keys when present.
        """
        row, col = self._resolve_address(address)
        result = self._bridge.call_json(
            "compute_get_hyperlink", self._sheet_id_json, row, col
        )
        if isinstance(result, dict):
            return result
        if isinstance(result, str) and result.startswith("{"):
            try:
                parsed = json.loads(result)
                if isinstance(parsed, dict):
                    return parsed
            except (json.JSONDecodeError, TypeError):
                pass
        return None

    def add(
        self,
        address: Union[str, Tuple[int, int]],
        link_or_url: Union[str, Dict[str, Any], None] = None,
        display: Optional[str] = None,
        tooltip: Optional[str] = None,
    ) -> MutationResult:
        """Add a hyperlink to a cell (alias for :meth:`set`).

        Parameters
        ----------
        address:
            A1 address or ``(row, col)`` tuple.
        link_or_url:
            Either a URL string or a dict with ``url``, ``display``, ``tooltip`` keys.
        display:
            Optional display text (used when *link_or_url* is a string).
        tooltip:
            Optional tooltip text (used when *link_or_url* is a string).
        """
        if isinstance(link_or_url, dict):
            url = link_or_url.get("url", "")
            display = link_or_url.get("display", display)
            tooltip = link_or_url.get("tooltip", tooltip)
        else:
            url = link_or_url or ""
        return self.set(address, url, display=display, tooltip=tooltip)

    def remove(self, address: Union[str, Tuple[int, int]]) -> MutationResult:
        """Remove the hyperlink from a cell.

        Parameters
        ----------
        address:
            A1 address or ``(row, col)`` tuple.
        """
        row, col = self._resolve_address(address)
        raw = self._bridge.call_json(
            "compute_remove_hyperlink", self._sheet_id_json, row, col
        )
        return deserialize_mutation_result(raw)

File: types/api/src/api/worksheet/hyperlinks.ts
/**
 * WorksheetHyperlinks — Sub-API for cell hyperlink operations.
 *
 * Provides methods to set, get, and remove hyperlinks on cells.
 */

/** Public readback for a worksheet hyperlink. */
export interface WorksheetHyperlink {
  /** A1-style cell address or range reference. */
  address: string;
  /** A1-style cell address or range reference. Alias for `address`. */
  ref: string;
  /** Hyperlink target URL or workbook location. */
  url: string;
  /** Optional display text stored with the hyperlink. */
  display?: string;
  /** Optional tooltip text stored with the hyperlink. */
  tooltip?: string;
}

/** Sub-API for hyperlink operations on a worksheet. */
export interface WorksheetHyperlinks {
  /**
   * Set a hyperlink on a cell.
   *
   * @param address - A1-style cell address (e.g. "A1")
   * @param url - Hyperlink URL
   */
  set(address: string, url: string): Promise<void>;
  /**
   * Set a hyperlink on a cell.
   *
   * @param row - Row index (0-based)
   * @param col - Column index (0-based)
   * @param url - Hyperlink URL
   */
  set(row: number, col: number, url: string): Promise<void>;

  /**
   * Get the hyperlink URL for a cell.
   *
   * @param address - A1-style cell address
   * @returns The hyperlink URL, or null if no hyperlink
   */
  get(address: string): Promise<string | null>;
  /**
   * Get the hyperlink URL for a cell.
   *
   * @param row - Row index (0-based)
   * @param col - Column index (0-based)
   * @returns The hyperlink URL, or null if no hyperlink
   */
  get(row: number, col: number): Promise<string | null>;

  /**
   * Check if a cell has a hyperlink.
   *
   * @param address - A1-style cell address
   * @returns True if the cell has a hyperlink
   */
  has(address: string): Promise<boolean>;
  /**
   * Check if a cell has a hyperlink.
   *
   * @param row - Row index (0-based)
   * @param col - Column index (0-based)
   * @returns True if the cell has a hyperlink
   */
  has(row: number, col: number): Promise<boolean>;

  /**
   * Remove a hyperlink from a cell.
   *
   * @param address - A1-style cell address
   */
  remove(address: string): Promise<void>;
  /**
   * Remove a hyperlink from a cell.
   *
   * @param row - Row index (0-based)
   * @param col - Column index (0-based)
   */
  remove(row: number, col: number): Promise<void>;

  /**
   * List all hyperlinks in the worksheet.
   *
   * @returns Array of hyperlink entries with cell address, URL, and metadata
   */
  list(): Promise<WorksheetHyperlink[]>;

  /**
   * Remove all hyperlinks from the worksheet.
   */
  clear(): Promise<void>;
}
